### PR TITLE
Implementation of InstanceAdmin::DeleteInstance call

### PR DESCRIPTION
**Pull request by manish-qlogic**
This commit includes:
    Implementing the integration test. The CreateInstance Call in integration test is a work in progress. 
    Implementing the sample snippet.
    Implementing the unit tests for "no errors", "unrecoverable error", and "no retry with recoverable errors" (because it is non-idempotent).
    Implementing the function itself.

### CODE REVIEW

File: bigtable/client/instance_admin_test.cc [hunk 1]
```diff
@@ -422,3 +422,51 @@ TEST_F(InstanceAdminTest, CreateInstancePollReturnsFailure) {
...
+                            "exceptions are disabled");
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+}
+
+/// @test Verify that recoverable error for DeleteInstance
+TEST_F(InstanceAdminTest, DeleteInstanceRecoverableError) {
+  using namespace ::testing;
+  bigtable::InstanceAdmin tested(client_);
+  EXPECT_CALL(*client_, DeleteInstance(_, _, _))
+      .Times(1)
```
Comment (coryan) [inline bigtable/client/instance_admin_test.cc]: I think `.Times(1)` and `WillOnce()` is redundant.  Remove the `Times(1)` ?

File: bigtable/examples/BUILD [hunk 2]
```diff
@@ -26,3 +26,9 @@ cc_binary(
     srcs=["bigtable_samples.cc"],
     deps=["//bigtable/client:bigtable_client"],
 )
+
+cc_binary(
+    name="bigtable_samples_instance_admin",
+    srcs=["bigtable_samples_instance_admin.cc"],
+    deps=["//bigtable/client:bigtable_client"],
+)
```
Comment (coryan) [inline bigtable/examples/BUILD]: Missing newline, please add.

File: bigtable/examples/bigtable_samples_instance_admin.cc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+//! [create instance]
+void CreateInstance(bigtable::InstanceAdmin& instance_admin,
+                    std::string const& project_id,
+                    std::string const& instance_id) {
+  // TODO(#418) implement tests and examples for CreateInstance
+}
+//! [create instance]
+
+//! [list instance]
```
Comment (coryan) [inline bigtable/examples/bigtable_samples_instance_admin.cc]: Please call the snippet `[list instances]` (plural) for consistency.

File: bigtable/examples/bigtable_samples_instance_admin.cc [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+//! [delete instance]
+void DeleteInstance(bigtable::InstanceAdmin& instance_admin,
+                    std::string const& project_id,
+                    std::string const& instance_id) {
+  instance_admin.DeleteInstance(instance_id);
+}
+//! [delete instance]
+
+//! [list cluster]
```
Comment (coryan) [inline bigtable/examples/bigtable_samples_instance_admin.cc]: s/list cluster/list clusters/

File: bigtable/examples/bigtable_samples_instance_admin.cc [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+#include "bigtable/client/instance_admin.h"
+#include "bigtable/client/instance_admin_client.h"
+
+namespace btproto = ::google::bigtable::admin::v2;
+
+namespace {
+
+//! [create instance]
+void CreateInstance(bigtable::InstanceAdmin& instance_admin,
```
Comment (coryan) [inline bigtable/examples/bigtable_samples_instance_admin.cc]: We fixed `bigtable::InstanceAdmin` to support copy constructors, just pass by value.

File: bigtable/examples/bigtable_samples_instance_admin.cc [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  //! [connect instance admin client]
+
+  // Connect to the Cloud Bigtable endpoint.
+  //! [connect instance admin]
+  bigtable::InstanceAdmin instance_admin(instance_admin_client);
+  //! [connect instance admin]
+
+  if (command == "create-instance") {
+    CreateInstance(instance_admin, project_id, instance_id);
+  } else if (command == "list-instance") {
```
Comment (coryan) [inline bigtable/examples/bigtable_samples_instance_admin.cc]: s/list-instance/list-instances/ (also in the example for "Usage").

File: bigtable/examples/bigtable_samples_instance_admin.cc [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    ListInstances(instance_admin);
+  } else if (command == "get-instance") {
+    GetInstance(instance_admin, project_id, instance_id);
+  } else if (command == "delete-instance") {
+    DeleteInstance(instance_admin, project_id, instance_id);
+  } else if (command == "list-cluster") {
```
Comment (coryan) [inline bigtable/examples/bigtable_samples_instance_admin.cc]: s/list-cluster/list-clusters/

File: bigtable/tests/instance_admin_integration_test.cc [hunk 8]
```diff
@@ -47,6 +47,18 @@ class InstanceAdminIntegrationTest : public ::testing::Test {
 bool UsingCloudBigtableEmulator() {
   return std::getenv("BIGTABLE_EMULATOR_HOST") != nullptr;
 }
+bool IsInstancePresent(
```
Comment (coryan) [inline bigtable/tests/instance_admin_integration_test.cc]: blank line please.

File: bigtable/tests/instance_admin_integration_test.cc [hunk 9]
```diff
@@ -47,6 +47,18 @@ class InstanceAdminIntegrationTest : public ::testing::Test {
...
+    std::vector<::google::bigtable::admin::v2::Instance> instances,
+    std::string instance_name) {
+  bool instance_present = false;
+  for (auto const& i : instances) {
+    if (instance_name.compare(i.name()) == 0) {
+      instance_present = true;
+      std::cout << "Instance Name " << i.name();
+    }
```
Comment (coryan) [inline bigtable/tests/instance_admin_integration_test.cc]: you can return as soon as the value becomes true, right?  You could also use `std::count()` or `std::find()`, but those are probably harder to read.

File: bigtable/tests/instance_admin_integration_test.cc [hunk 10]
```diff
@@ -47,6 +47,18 @@ class InstanceAdminIntegrationTest : public ::testing::Test {
 bool UsingCloudBigtableEmulator() {
   return std::getenv("BIGTABLE_EMULATOR_HOST") != nullptr;
 }
+bool IsInstancePresent(
+    std::vector<::google::bigtable::admin::v2::Instance> instances,
+    std::string instance_name) {
+  bool instance_present = false;
+  for (auto const& i : instances) {
+    if (instance_name.compare(i.name()) == 0) {
```
Comment (coryan) [inline bigtable/tests/instance_admin_integration_test.cc]: I think `instance_name == i.name()` is more natural in this case.

File: bigtable/tests/instance_admin_integration_test.cc [hunk 11]
```diff
@@ -47,6 +47,18 @@ class InstanceAdminIntegrationTest : public ::testing::Test {
 bool UsingCloudBigtableEmulator() {
   return std::getenv("BIGTABLE_EMULATOR_HOST") != nullptr;
 }
+bool IsInstancePresent(
+    std::vector<::google::bigtable::admin::v2::Instance> instances,
+    std::string instance_name) {
```
Comment (coryan) [inline bigtable/tests/instance_admin_integration_test.cc]: clang-tidy will probably recommend passing by const reference in this case.

File: bigtable/tests/instance_admin_integration_test.cc [hunk 12]
```diff
@@ -62,6 +74,34 @@ TEST_F(InstanceAdminIntegrationTest, ListInstancesTest) {
   }
 }
 
+/// @test Verify that InstanceAdmin::DeleteInstances works as expected.
+TEST_F(InstanceAdminIntegrationTest, DeleteInstancesTest) {
+  // The emulator does not support instance operations.
+  if (UsingCloudBigtableEmulator()) {
+    return;
+  }
+  std::string instance_id = "DeleteInstancesTest_instance";
```
Comment (coryan) [inline bigtable/tests/instance_admin_integration_test.cc]: no action needed here, but what we should do is create an random instance name, that way the opportunity for collision is really small.  We do something like that in the benchmarks (for table names).

File: bigtable/tests/instance_admin_integration_test.cc [hunk 13]
```diff
@@ -62,6 +74,34 @@ TEST_F(InstanceAdminIntegrationTest, ListInstancesTest) {
...
+  bigtable::InstanceId instance(instance_id);
+  bigtable::DisplayName diplay_name(instance_id);
+  EXPECT_FALSE(
+      IsInstancePresent(instance_admin_->ListInstances(), instance_id));
+
+  std::vector<std::pair<std::string, bigtable::ClusterConfig>> clusters;
+  clusters.push_back(std::make_pair(
+      "sample-cluster",
+      bigtable::ClusterConfig(
+          "sample-cluster", 1234,
```
Comment (coryan) [inline bigtable/tests/instance_admin_integration_test.cc]: What is `1234` here?

File: bigtable/tests/instance_admin_integration_test.cc [hunk 14]
```diff
@@ -62,6 +74,34 @@ TEST_F(InstanceAdminIntegrationTest, ListInstancesTest) {
...
+          google::bigtable::admin::v2::StorageType::STORAGE_TYPE_UNSPECIFIED)));
```
Comment (coryan) [inline bigtable/tests/instance_admin_integration_test.cc]: I think you have to set the storage type (use HDD) or the server will fail the request.

File: bigtable/tests/instance_admin_integration_test.cc (hunk 13 above)
Comment (coryan) [reply bigtable/tests/instance_admin_integration_test.cc]: Ah, the number of nodes I think.  Please change this to create a development instance, production instances require at least 3 nodes and they are expensive (and slower) to create.

File: bigtable/client/internal/instance_admin_test.cc [hunk 15]
```diff
@@ -250,3 +250,46 @@ TEST_F(InstanceAdminTest, GetInstance) {
...
+  tested.DeleteInstance("the-instance", status);
+  EXPECT_TRUE(status.ok());
+}
+
+/// @test Verify unrecoverable error for DeleteInstance
+TEST_F(InstanceAdminTest, DeleteInstanceUnrecoverableError) {
+  using namespace ::testing;
+  bigtable::noex::InstanceAdmin tested(client_);
+  EXPECT_CALL(*client_, DeleteInstance(_, _, _))
+      .WillRepeatedly(
```
Comment (coryan) [inline bigtable/client/internal/instance_admin_test.cc]: I think this should also be a `.WillOnce`.

File: bigtable/client/instance_admin_test.cc (hunk 1 above)
Comment (manish-qlogic) [reply bigtable/client/instance_admin_test.cc]: removed Times(1), but WillOnce causes failure if run with -DGOOGLE_CLOUD_CPP_ENABLE_CXX_EXCEPTIONS=no.
The error is
` RUN      ] InstanceAdminTest.DeleteInstanceRecoverableError
/home/manish/git/google-cloud-cpp/bigtable/client/instance_admin_test.cc:466: Failure
Actual function call count doesn't match EXPECT_CALL(*client_, DeleteInstance(_, _, _))...
         Expected: to be called once
           Actual: never called - unsatisfied and active
[  FAILED  ] InstanceAdminTest.DeleteInstanceRecoverableError (150 ms)
`  

I am not sure of the reason behind it, the work around is to use WillRepeatedly instead.

File: bigtable/examples/BUILD (hunk 2 above)
Comment (manish-qlogic) [reply bigtable/examples/BUILD]: added new line

File: bigtable/examples/bigtable_samples_instance_admin.cc (hunk 3 above)
Comment (manish-qlogic) [reply bigtable/examples/bigtable_samples_instance_admin.cc]: done.

File: bigtable/examples/bigtable_samples_instance_admin.cc (hunk 4 above)
Comment (manish-qlogic) [reply bigtable/examples/bigtable_samples_instance_admin.cc]: done.

File: bigtable/examples/bigtable_samples_instance_admin.cc (hunk 5 above)
Comment (manish-qlogic) [reply bigtable/examples/bigtable_samples_instance_admin.cc]: done it for all the calls.

File: bigtable/examples/bigtable_samples_instance_admin.cc (hunk 6 above)
Comment (manish-qlogic) [reply bigtable/examples/bigtable_samples_instance_admin.cc]: done.

File: bigtable/examples/bigtable_samples_instance_admin.cc (hunk 7 above)
Comment (manish-qlogic) [reply bigtable/examples/bigtable_samples_instance_admin.cc]: fixed.

File: bigtable/tests/instance_admin_integration_test.cc (hunk 8 above)
Comment (manish-qlogic) [reply bigtable/tests/instance_admin_integration_test.cc]: inserted

File: bigtable/tests/instance_admin_integration_test.cc (hunk 9 above)
Comment (manish-qlogic) [reply bigtable/tests/instance_admin_integration_test.cc]: fixed by breaking the loop on match.

File: bigtable/tests/instance_admin_integration_test.cc (hunk 10 above)
Comment (manish-qlogic) [reply bigtable/tests/instance_admin_integration_test.cc]: fixed.

File: bigtable/tests/instance_admin_integration_test.cc (hunk 11 above)
Comment (manish-qlogic) [reply bigtable/tests/instance_admin_integration_test.cc]: fixed

File: bigtable/tests/instance_admin_integration_test.cc (hunk 13 above)
Comment (manish-qlogic) [reply bigtable/tests/instance_admin_integration_test.cc]: changed it to 1.

File: bigtable/client/instance_admin_test.cc (hunk 1 above)
Comment (coryan) [reply bigtable/client/instance_admin_test.cc]: Ahh, yes, the usual problem with `EXPECT_DEATH_.*`, it runs in a separate process, so the call is not counted (or visible) to the parent process.  Thanks for checking.

File: bigtable/tests/instance_admin_integration_test.cc (hunk 14 above)
Comment (manish-qlogic) [reply bigtable/tests/instance_admin_integration_test.cc]: changed it HDD.

File: bigtable/client/internal/instance_admin_test.cc (hunk 15 above)
Comment (manish-qlogic) [reply bigtable/client/internal/instance_admin_test.cc]: changed to `.WillOnce`

**Comment by manish-qlogic** [comment]
PTAL

**Pull request merged**